server/src/main/java/org/apache/iotdb/db/metadata/template/TemplateManager.java

```diff
@@ -0,0 +1,135 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iotdb.db.metadata.template;
+
+import org.apache.iotdb.db.exception.metadata.DuplicatedTemplateException;
+import org.apache.iotdb.db.exception.metadata.MetadataException;
+import org.apache.iotdb.db.exception.metadata.PathAlreadyExistException;
+import org.apache.iotdb.db.exception.metadata.UndefinedTemplateException;
+import org.apache.iotdb.db.metadata.mnode.IMNode;
+import org.apache.iotdb.db.qp.physical.crud.CreateTemplatePlan;
+import org.apache.iotdb.tsfile.utils.Pair;
+import org.apache.iotdb.tsfile.write.schema.IMeasurementSchema;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class TemplateManager {
+
+  // template name -> template
+  private Map<String, Template> templateMap = new ConcurrentHashMap<>();
+
+  private static class TemplateManagerHolder {
+
+    private TemplateManagerHolder() {
+      // allowed to do nothing
+    }
+
+    private static final TemplateManager INSTANCE = new TemplateManager();
+  }
+
+  public static TemplateManager getInstance() {
+    return TemplateManagerHolder.INSTANCE;
+  }
+
+  private TemplateManager() {}
+
+  public void createDeviceTemplate(CreateTemplatePlan plan) throws MetadataException {
+    Template template = new Template(plan);
+    if (templateMap.putIfAbsent(plan.getName(), template) != null) {
+      // already have template
+      throw new MetadataException("Duplicated template name: " + plan.getName());
+    }
+  }
+
+  public Template getTemplate(String templateName) throws UndefinedTemplateException {
+    Template template = templateMap.get(templateName);
+    if (template == null) {
+      throw new UndefinedTemplateException(templateName);
+    }
+    return template;
+  }
+
+  public void setDeviceTemplate(Template template, Pair<IMNode, Template> node)
+      throws MetadataException {
+
+    if (node.left.getDeviceTemplate() != null) {
+      if (node.left.getDeviceTemplate().equals(template)) {
+        throw new DuplicatedTemplateException(template.getName());
+      } else {
+        throw new MetadataException("Specified node already has template");
+      }
+    }
+
+    if (!isTemplateCompatible(node.right, template)) {
+      throw new MetadataException("Incompatible template");
+    }
+
+    checkIsTemplateAndMNodeCompatible(template, node.left);
+
+    node.left.setDeviceTemplate(template);
+  }
+
+  public boolean isTemplateCompatible(Template upper, Template current) {
+    if (upper == null) {
+      return true;
```

Comment:
upper == null, return true?

Reply:
If the upperTemplate is null, then the current template to be set  will have no conflict with current mtree. The rule is that the lower level template should contains all schema defined in upper template.